labelCloud/io/labels/kitti.py

```diff
@@ -20,72 +23,130 @@ def __init__(
-                        0, 0, rel2abs_rotation(-float(line_elements[14]) + math.pi / 2)
-                    )
-                else:
-                    bbox.set_rotations(0, 0, rel2abs_rotation(float(line_elements[14])))
-                bbox.set_classname(line_elements[0])
-                labels.append(bbox)
+                rotation = -float(meta["rotation_y"]) + math.pi / 2 if self.transformed else float(meta["rotation_y"])
+                bbox.set_rotations(0, 0, rel2abs_rotation(rotation))
+                bbox.set_classname(meta["type"])
+                bboxes.append(bbox)
             logging.info("Imported %s labels from %s." % (len(label_lines), label_path))
-        return labels
+        return bboxes
 
     def export_labels(self, bboxes: List[BBox], pcd_path: Path) -> None:
         data = str()
 
         # Labels
-        for bbox in bboxes:
+        for i, bbox in enumerate(bboxes):
             obj_type = bbox.get_classname()
             centroid = bbox.get_center()
-            if self.transformed:
-                centroid = (-centroid[1], centroid[2] + 2.3, centroid[0])
-            location = " ".join([str(self.round_dec(v)) for v in centroid])
             dimensions = bbox.get_dimensions()
             if self.transformed:
-                dimensions = (dimensions[2], dimensions[1], dimensions[0])
+                centroid = (
+                    centroid[0],
+                    centroid[1],
+                    centroid[2] - dimensions[2] / 2,
+                )  # centroid in KITTI located on bottom face of bbox
+                dimensions = dimensions[2], dimensions[1], dimensions[0]
+                xyz1 = np.insert(np.asarray(centroid), 3, values=[1])
+                xyz1 = self.T_v2c @ xyz1
+                centroid = tuple([float(n) for n in xyz1[:-1]])
+            location_str = " ".join([str(self.round_dec(v)) for v in centroid])
             dimensions_str = " ".join([str(self.round_dec(v)) for v in dimensions])
-            rotation_z = bbox.get_z_rotation()
-            if self.transformed:
-                rotation_y = self.round_dec(
-                    -(abs2rel_rotation(rotation_z) - math.pi / 2)
-                )
-            else:
-                rotation_y = self.round_dec(abs2rel_rotation(rotation_z))
-
-            data += (
-                " ".join(
-                    [
-                        obj_type,
-                        "0 0 0 0 0 0 0",
-                        dimensions_str,
-                        location,
-                        str(rotation_y),
-                    ]
-                )
-                + "\n"
-            )
+            rotation = bbox.get_z_rotation()
+            rotation = abs2rel_rotation(rotation)
+            rotation = -(rotation - math.pi / 2) if self.transformed else rotation
+            rotation = str(self.round_dec(rotation))
+
+            out_str = list(self.bboxes_meta[i].values())
+            if obj_type != "DontCare":
```

Comment:
Why are you explicitly omitting the "DontCare" objects?

They could change as well, or?

Reply:
@Divelix Would be great if you can explain here why or otherwise remove the condition.

Reply:
As stated [here](https://github.com/bostondiditeam/kitti/blob/master/resources/devkit_object/readme.txt#L65), "DontCare" class is specific to KITTI and intended to help in filtering out hard false positives **on images** (not pointclouds) during training. If you look at any label file with "DontCare" objects, you'll see that it contains only **image 2d bbox** values, which labelCloud don't use.
For example, `000005` label from .zip I sent you:
```
Pedestrian 0.00 0 1.94 330.06 178.74 360.77 238.64 1.87 0.96 0.65 -8.50 2.07 23.02 1.59
DontCare -1 -1 -10 606.64 170.67 621.06 184.34 -1 -1 -1 -1000 -1000 -1000 -10
DontCare -1 -1 -10 606.00 170.91 621.35 184.28 -1 -1 -1 -1000 -1000 -1000 -10
DontCare -1 -1 -10 605.68 171.21 620.77 184.34 -1 -1 -1 -1000 -1000 -1000 -10
DontCare -1 -1 -10 566.39 168.89 585.07 184.56 -1 -1 -1 -1000 -1000 -1000 -10
```

Reply:
I see.

Then they probably should also be skipped, when reading labels from file, or?

Currently, labelCloud would then remove them from all labels, when going through the point clouds.

Do you think that is the desired effect?

Another option might be to not load but cache them and then write them as is again, when writing the labels.

Reply:
It work fine right now, I think